File: tests/wasm/test_mass.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
const { describe, it } = test;


const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);


import * as talib from '../../pkg/vector_ta.js';

describe('Mass Index', () => {
    let testData;

    test.before(() => {
        testData = loadTestData();
    });

    it('should calculate mass with partial parameters - mirrors check_mass_partial_params', () => {
        const result = talib.mass_js(testData.high, testData.low, 5);
        assert.strictEqual(result.length, testData.high.length);
    });

    it('should match expected accuracy - mirrors check_mass_accuracy', () => {
        const expected = EXPECTED_OUTPUTS.mass;
        const result = talib.mass_js(
            testData.high,
            testData.low,
            expected.defaultParams.period
        );

        assert.strictEqual(result.length, testData.high.length);


        const lastFive = result.slice(-5);
        assertArrayClose(
            lastFive,
            expected.last5Values,
            1e-7,
            "Mass Index last 5 values mismatch"
        );
    });

    it('should handle default candles - mirrors check_mass_default_candles', () => {
        const result = talib.mass_js(testData.high, testData.low, 5);
        assert.strictEqual(result.length, testData.high.length);
    });

    it('should handle zero period - mirrors check_mass_zero_period', () => {
        assert.throws(() => {
            talib.mass_js([10, 15, 20], [5, 10, 12], 0);
        }, /Invalid period/, 'Should throw for zero period');
    });

    it('should handle period exceeds length - mirrors check_mass_period_exceeds_length', () => {
        assert.throws(() => {
            talib.mass_js([10, 15, 20], [5, 10, 12], 10);
        }, /Invalid period/, 'Should throw when period > data length');
    });

    it('should handle very small dataset - mirrors check_mass_very_small_data_set', () => {
        assert.throws(() => {
            talib.mass_js([10.0], [5.0], 5);
        }, /Invalid period|Not enough valid data/, 'Should throw for insufficient data');
    });

    it('should handle empty data - mirrors check_mass_empty_input', () => {
        assert.throws(() => {
            talib.mass_js([], [], 5);
        }, /empty/i, 'Should throw for empty data');
    });

    it('should handle different length high/low - mirrors check_mass_different_length_hl', () => {
        assert.throws(() => {
            talib.mass_js([1, 2, 3], [1, 2], 2);
        }, /High and low/, 'Should throw for mismatched high/low lengths');
    });

    it('should handle NaN values correctly - mirrors check_mass_nan_handling', () => {
        const expected = EXPECTED_OUTPUTS.mass;
        const result = talib.mass_js(testData.high, testData.low, 5);

        assert.strictEqual(result.length, testData.high.length);


        const warmup = expected.warmupPeriod;
        for (let i = 0; i < warmup; i++) {
            assert.ok(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
        }


        assert.ok(!isNaN(result[warmup]), `Expected valid value at index ${warmup} (first valid index)`);


        if (result.length > 240) {
            for (let i = 240; i < result.length; i++) {
                assert.ok(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
            }
        }
    });

    it('should handle all NaN values - mirrors check_mass_all_nan', () => {
        const allNaN = new Array(100).fill(NaN);

        assert.throws(() => {
            talib.mass_js(allNaN, allNaN, 5);
        }, /All values are NaN/, 'Should throw for all NaN input');
    });

    it('should handle batch operations with metadata', () => {
        const high = testData.high.slice(0, 100);
        const low = testData.low.slice(0, 100);

        const config = {
            period_range: [5, 15, 5]
        };

        const result = talib.mass_batch(high, low, config);


        assert.strictEqual(result.rows, 3, 'Should have 3 rows');
        assert.strictEqual(result.cols, 100, 'Should have 100 columns');
        assert.strictEqual(result.values.length, 300, 'Should have 300 values total');
        assert.strictEqual(result.combos.length, 3, 'Should have 3 parameter combinations');


        assert.strictEqual(result.combos[0].period, 5, 'First combo period should be 5');
        assert.strictEqual(result.combos[1].period, 10, 'Second combo period should be 10');
        assert.strictEqual(result.combos[2].period, 15, 'Third combo period should be 15');


        const periods = [5, 10, 15];
        periods.forEach((period, row) => {
            const single = talib.mass_js(high, low, period);
            const rowStart = row * 100;
            const rowEnd = rowStart + 100;
            const batchRow = result.values.slice(rowStart, rowEnd);

            assertArrayClose(
                batchRow,
                single,
                1e-10,
                `Batch row ${row} (period=${period}) should match single calculation`
            );
        });
    });

    it('should handle batch parameter sweep', () => {
        const high = testData.high.slice(0, 30);
        const low = testData.low.slice(0, 30);

        const config = {
            period_range: [3, 7, 2]
        };

        const result = talib.mass_batch(high, low, config);

        assert.strictEqual(result.rows, 3, 'Should have 3 rows for sweep');
        assert.strictEqual(result.combos.length, 3, 'Should have 3 combinations');
        assert.deepStrictEqual(
            result.combos.map(c => c.period),
            [3, 5, 7],
            'Should have correct period values'
        );


        result.combos.forEach((combo, row) => {
            const warmup = 16 + combo.period - 1;
            const rowStart = row * 30;
            const rowData = result.values.slice(rowStart, rowStart + 30);


            for (let i = 0; i < Math.min(warmup, 30); i++) {
                assert.ok(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${combo.period}`);
            }


            if (warmup < 30) {
                assert.ok(!isNaN(rowData[warmup]), `Expected valid value after warmup for period ${combo.period}`);
            }
        });
    });

    it('should handle zero-copy API - in-place operation', () => {
        const high = testData.high.slice(0, 100);
        const low = testData.low.slice(0, 100);
        const len = high.length;


        const outPtr = talib.mass_alloc(len);

        try {

            const expected = talib.mass_js(high, low, 5);


            const highPtr = talib.mass_alloc(len);
            const lowPtr = talib.mass_alloc(len);


            const wasmMemory = new Float64Array(talib.__wasm.memory.buffer);
            wasmMemory.set(high, highPtr / 8);
            wasmMemory.set(low, lowPtr / 8);


            talib.mass_into(highPtr, lowPtr, outPtr, len, 5);


            const result = new Float64Array(talib.__wasm.memory.buffer, outPtr, len);


            assertArrayClose(
                Array.from(result),
                expected,
                1e-10,
                "Fast API should match safe API"
            );


            talib.mass_into(highPtr, lowPtr, highPtr, len, 5);
            const inPlaceResult = new Float64Array(talib.__wasm.memory.buffer, highPtr, len);

            assertArrayClose(
                Array.from(inPlaceResult),
                expected,
                1e-10,
                "In-place operation should match expected"
            );


            talib.mass_free(highPtr, len);
            talib.mass_free(lowPtr, len);

        } finally {

            talib.mass_free(outPtr, len);
        }
    });

    it('should handle zero-copy error cases', () => {

        assert.throws(() => {
            talib.mass_into(0, 0, 0, 10, 5);
        }, /null pointer/, 'Should throw for null pointer');


        const ptr = talib.mass_alloc(10);
        try {

            assert.throws(() => {
                talib.mass_into(ptr, ptr, ptr, 10, 0);
            }, /Invalid period/, 'Should throw for invalid period');


            assert.throws(() => {
                talib.mass_into(ptr, ptr, ptr, 10, 20);
            }, /Invalid period/, 'Should throw when period exceeds length');
        } finally {
            talib.mass_free(ptr, 10);
        }
    });

    it('should handle zero-copy memory management', () => {

        const sizes = [100, 1000, 10000];

        for (const size of sizes) {
            const ptr = talib.mass_alloc(size);
            assert.ok(ptr !== 0, `Failed to allocate ${size} elements`);


            const memView = new Float64Array(talib.__wasm.memory.buffer, ptr, size);
            for (let i = 0; i < Math.min(10, size); i++) {
                memView[i] = i * 1.5;
            }


            for (let i = 0; i < Math.min(10, size); i++) {
                assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
            }


            talib.mass_free(ptr, size);
        }
    });

    it('should handle batch with fast API', () => {
        const high = testData.high.slice(0, 50);
        const low = testData.low.slice(0, 50);
        const len = high.length;


        const periodStart = 5, periodEnd = 10, periodStep = 5;
        const expectedRows = 2;


        const highPtr = talib.mass_alloc(len);
        const lowPtr = talib.mass_alloc(len);
        const outPtr = talib.mass_alloc(len * expectedRows);

        try {

            const wasmMemory = new Float64Array(talib.__wasm.memory.buffer);
            wasmMemory.set(high, highPtr / 8);
            wasmMemory.set(low, lowPtr / 8);


            const rows = talib.mass_batch_into(
                highPtr, lowPtr, outPtr, len,
                periodStart, periodEnd, periodStep
            );

            assert.strictEqual(rows, expectedRows, 'Should return correct number of rows');


            const results = new Float64Array(talib.__wasm.memory.buffer, outPtr, len * rows);


            const periods = [5, 10];
            periods.forEach((period, row) => {
                const expected = talib.mass_js(high, low, period);
                const rowStart = row * len;
                const rowData = Array.from(results.slice(rowStart, rowStart + len));

                assertArrayClose(
                    rowData,
                    expected,
                    1e-10,
                    `Batch fast API row ${row} (period=${period}) should match`
                );
            });

        } finally {
            talib.mass_free(highPtr, len);
            talib.mass_free(lowPtr, len);
            talib.mass_free(outPtr, len * expectedRows);
        }
    });

    it('should support WASM streaming API', () => {

        const stream = new talib.MassStreamWasm(5);


        const highData = testData.high.slice(0, 50);
        const lowData = testData.low.slice(0, 50);


        const batchResult = talib.mass_js(highData, lowData, 5);


        const streamResults = [];
        for (let i = 0; i < highData.length; i++) {
            const result = stream.update(highData[i], lowData[i]);
            streamResults.push(result !== null && result !== undefined ? result : NaN);
        }


        const warmup = EXPECTED_OUTPUTS.mass.warmupPeriod;
        for (let i = warmup; i < streamResults.length; i++) {
            assertClose(
                streamResults[i],
                batchResult[i],
                1e-10,
                `Streaming result at index ${i} should match batch result`
            );
        }
    });

    it('should handle invalid period in streaming', () => {
        assert.throws(() => {
            new talib.MassStreamWasm(0);
        }, /Invalid period/, 'Should throw for invalid period in streaming');
    });

    it('should verify SIMD128 consistency', () => {


        const testCases = [
            { size: 25, period: 3 },
            { size: 100, period: 5 },
            { size: 1000, period: 10 },
            { size: 5000, period: 20 }
        ];

        for (const testCase of testCases) {

            const high = new Float64Array(testCase.size);
            const low = new Float64Array(testCase.size);
            for (let i = 0; i < testCase.size; i++) {
                const base = Math.sin(i * 0.1) * 100 + 1000;
                const range = Math.abs(Math.cos(i * 0.05)) * 10 + 5;
                high[i] = base + range;
                low[i] = base - range;
            }

            const result = talib.mass_js(high, low, testCase.period);


            assert.strictEqual(result.length, high.length);


            const warmup = 16 + testCase.period - 1;
            for (let i = 0; i < Math.min(warmup, result.length); i++) {
                assert.ok(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}, period=${testCase.period}`);
            }


            let sumAfterWarmup = 0;
            let countAfterWarmup = 0;
            for (let i = warmup; i < result.length; i++) {
                assert.ok(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}, period=${testCase.period}`);
                sumAfterWarmup += result[i];
                countAfterWarmup++;
            }


            if (countAfterWarmup > 0) {
                const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
                assert.ok(avgAfterWarmup > 0, `Average Mass Index should be positive, got ${avgAfterWarmup}`);
                assert.ok(avgAfterWarmup < testCase.period * 2, `Average Mass Index seems too high: ${avgAfterWarmup}`);
            }
        }
    });

    it('should handle batch edge cases', () => {
        const high = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
                                        11, 12, 13, 14, 15, 16, 17, 18, 19, 20]);
        const low = high.map(h => h * 0.9);


        const singleBatch = talib.mass_batch(high, low, {
            period_range: [3, 3, 1]
        });

        assert.strictEqual(singleBatch.values.length, 20);
        assert.strictEqual(singleBatch.combos.length, 1);
        assert.strictEqual(singleBatch.combos[0].period, 3);


        const largeBatch = talib.mass_batch(high, low, {
            period_range: [3, 5, 10]
        });


        assert.strictEqual(largeBatch.values.length, 20);
        assert.strictEqual(largeBatch.combos.length, 1);
        assert.strictEqual(largeBatch.combos[0].period, 3);
    });

    test.after(() => {
        console.log('Mass Index WASM tests completed');
    });
});

File: tests/python/test_mass.py
"""
Python binding tests for Mass Index indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestMass:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_mass_partial_params(self, test_data):
        """Test Mass Index with partial parameters - mirrors check_mass_partial_params"""
        high = test_data['high']
        low = test_data['low']
        
        # Test with default period (5)
        result = ta_indicators.mass(high, low, 5)
        assert len(result) == len(high)
    
    def test_mass_accuracy(self, test_data):
        """Test Mass Index matches expected values from Rust tests - mirrors check_mass_accuracy"""
        high = test_data['high']
        low = test_data['low']
        expected = EXPECTED_OUTPUTS['mass']
        
        result = ta_indicators.mass(
            high,
            low,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(high)
        
        # Check last 5 values match expected
        # Rust test uses absolute tolerance of 1e-7; match or tighten (no looser than Rust)
        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0.0,
            atol=1e-7,
            msg="Mass Index last 5 values mismatch"
        )
        
        # Compare full output with Rust
        compare_with_rust('mass', result, 'high,low', expected['default_params'])
    
    def test_mass_default_candles(self, test_data):
        """Test Mass Index with default parameters - mirrors check_mass_default_candles"""
        high = test_data['high']
        low = test_data['low']
        
        # Default period is 5
        result = ta_indicators.mass(high, low, 5)
        assert len(result) == len(high)
    
    def test_mass_zero_period(self):
        """Test Mass Index fails with zero period - mirrors check_mass_zero_period"""
        high_data = np.array([10.0, 15.0, 20.0])
        low_data = np.array([5.0, 10.0, 12.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mass(high_data, low_data, period=0)
    
    def test_mass_period_exceeds_length(self):
        """Test Mass Index fails when period exceeds data length - mirrors check_mass_period_exceeds_length"""
        high_data = np.array([10.0, 15.0, 20.0])
        low_data = np.array([5.0, 10.0, 12.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mass(high_data, low_data, period=10)
    
    def test_mass_very_small_dataset(self):
        """Test Mass Index fails with insufficient data - mirrors check_mass_very_small_data_set"""
        high_data = np.array([10.0])
        low_data = np.array([5.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.mass(high_data, low_data, period=5)
    
    def test_mass_empty_input(self):
        """Test Mass Index fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.mass(empty, empty, period=5)
    
    def test_mass_different_length_hl(self):
        """Test Mass Index fails when high and low have different lengths"""
        high_data = np.array([10.0, 15.0, 20.0])
        low_data = np.array([5.0, 10.0])
        
        with pytest.raises(ValueError, match="High and low"):
            ta_indicators.mass(high_data, low_data, period=2)
    
    def test_mass_reinput(self, test_data):
        """Test Mass Index applied twice (re-input) - mirrors check_mass_reinput"""
        high = test_data['high']
        low = test_data['low']
        
        # First pass
        first_result = ta_indicators.mass(high, low, period=5)
        assert len(first_result) == len(high)
        
        # Second pass - apply Mass to Mass output
        second_result = ta_indicators.mass(first_result, first_result, period=5)
        assert len(second_result) == len(first_result)
    
    def test_mass_nan_handling(self, test_data):
        """Test Mass Index handles NaN values correctly - mirrors check_mass_nan_handling"""
        high = test_data['high']
        low = test_data['low']
        expected = EXPECTED_OUTPUTS['mass']
        
        result = ta_indicators.mass(high, low, period=5)
        assert len(result) == len(high)
        
        # Check warmup period has NaN values (first 20 values for period=5)
        warmup = expected['warmup_period']
        assert np.all(np.isnan(result[:warmup])), f"Expected NaN in first {warmup} values (warmup period)"
        
        # After warmup, values should be valid
        assert not np.isnan(result[warmup]), f"Expected valid value at index {warmup} (first valid index)"
        
        # After index 240, no NaN values should exist
        if len(result) > 240:
            for i in range(240, len(result)):
                assert not np.isnan(result[i]), f"Found unexpected NaN at index {i}"
    
    def test_mass_batch_single_parameter(self, test_data):
        """Test Mass batch with single parameter combination"""
        high = test_data['high']
        low = test_data['low']
        
        # Single period
        batch_result = ta_indicators.mass_batch(
            high, 
            low,
            period_range=(5, 5, 1)
        )
        
        # Should match single calculation
        single_result = ta_indicators.mass(high, low, 5)
        
        assert batch_result['values'].shape == (1, len(high))
        assert_close(
            batch_result['values'][0], 
            single_result,
            rtol=1e-10,
            msg="Batch vs single calculation mismatch"
        )
    
    def test_mass_batch_multiple_periods(self, test_data):
        """Test Mass batch with multiple period values"""
        high = test_data['high'][:100]  # Use smaller dataset for speed
        low = test_data['low'][:100]
        
        # Multiple periods: 5, 10, 15
        batch_result = ta_indicators.mass_batch(
            high,
            low,
            period_range=(5, 15, 5)
        )
        
        # Should have 3 rows * 100 cols
        assert batch_result['values'].shape == (3, 100)
        assert len(batch_result['periods']) == 3
        
        # Verify each row matches individual calculation
        periods = [5, 10, 15]
        for i, period in enumerate(periods):
            single_result = ta_indicators.mass(high, low, period)
            assert_close(
                batch_result['values'][i], 
                single_result, 
                rtol=1e-10,
                msg=f"Period {period} mismatch"
            )
    
    def test_mass_batch_edge_cases(self, test_data):
        """Test edge cases for batch processing"""
        # Mass needs at least 16 + period - 1 data points
        # For period=3, that's 18 points minimum, so use 20 to be safe
        high = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 
                        11, 12, 13, 14, 15, 16, 17, 18, 19, 20], dtype=np.float64)
        low = high * 0.9
        
        # Single value sweep
        single_batch = ta_indicators.mass_batch(
            high,
            low,
            period_range=(3, 3, 1)
        )
        
        assert single_batch['values'].shape == (1, 20)
        assert len(single_batch['periods']) == 1
        
        # Step larger than range
        large_batch = ta_indicators.mass_batch(
            high,
            low,
            period_range=(3, 5, 10)  # Step larger than range
        )
        
        # Should only have period=3
        assert large_batch['values'].shape == (1, 20)
        assert len(large_batch['periods']) == 1
    
    def test_mass_streaming(self):
        """Test Mass streaming functionality"""
        # Create a streaming instance
        stream = ta_indicators.MassStream(period=5)
        
        # Test data
        high_data = [10.0, 15.0, 20.0, 25.0, 30.0, 35.0, 40.0, 45.0, 50.0, 55.0,
                     60.0, 65.0, 70.0, 75.0, 80.0, 85.0, 90.0, 95.0, 100.0, 105.0,
                     110.0]
        low_data = [h * 0.9 for h in high_data]
        
        results = []
        for h, l in zip(high_data, low_data):
            result = stream.update(h, l)
            results.append(result)
        
        # First 20 values should be None (warmup period = 16 + period - 1 = 20)
        for i in range(20):
            assert results[i] is None, f"Expected None at index {i}, got {results[i]}"
        
        # After warmup, should have values
        assert results[20] is not None
        assert isinstance(results[20], float)
    
    def test_mass_kernel_parameter(self, test_data):
        """Test Mass with different kernel parameters"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        
        # Test with auto kernel (default)
        result_auto = ta_indicators.mass(high, low, 5)
        
        # Test with scalar kernel
        result_scalar = ta_indicators.mass(high, low, 5, kernel='scalar')
        
        # Results should be very close
        assert_close(
            result_auto,
            result_scalar,
            rtol=1e-10,
            msg="Auto vs Scalar kernel mismatch"
        )
    
    def test_mass_all_nan_input(self):
        """Test Mass Index with all NaN values - mirrors ALMA test coverage"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.mass(all_nan, all_nan, period=5)
    
    def test_mass_batch_metadata(self, test_data):
        """Test Mass batch includes correct metadata - mirrors ALMA batch tests"""
        high = test_data['high'][:50]  # Use smaller dataset
        low = test_data['low'][:50]
        
        # Test with multiple periods
        batch_result = ta_indicators.mass_batch(
            high,
            low,
            period_range=(5, 10, 5)  # periods: 5, 10
        )
        
        # Verify metadata structure
        assert 'values' in batch_result
        assert 'periods' in batch_result
        assert batch_result['values'].shape == (2, 50)
        assert len(batch_result['periods']) == 2
        assert list(batch_result['periods']) == [5, 10]
        
        # Verify each row has correct warmup
        for i, period in enumerate([5, 10]):
            row = batch_result['values'][i]
            warmup = 16 + period - 1
            # Check warmup period has NaN
            assert np.all(np.isnan(row[:warmup])), f"Expected NaN in warmup for period {period}"
            # Check after warmup has values
            if warmup < len(row):
                assert not np.isnan(row[warmup]), f"Expected valid value after warmup for period {period}"
    
    def test_mass_batch_parameter_sweep(self, test_data):
        """Test Mass batch with full parameter sweep"""
        high = test_data['high'][:30]  # Small dataset for testing
        low = test_data['low'][:30]
        
        # Test sweep with step
        batch_result = ta_indicators.mass_batch(
            high,
            low,
            period_range=(3, 7, 2)  # periods: 3, 5, 7
        )
        
        assert batch_result['values'].shape == (3, 30)
        assert list(batch_result['periods']) == [3, 5, 7]
        
        # Verify each row matches single calculation
        for i, period in enumerate([3, 5, 7]):
            single = ta_indicators.mass(high, low, period)
            assert_close(
                batch_result['values'][i],
                single,
                rtol=1e-10,
                msg=f"Batch row {i} (period={period}) doesn't match single calculation"
            )
